internal/graph/check_test.go

```diff
@@ -581,3 +582,150 @@ condition condition1(param1: string) {
+		}
+
+		resp, err := union(ctx, concurrencyLimit, depthExceededHandler, falseHandler)
+		require.ErrorIs(t, ErrResolutionDepthExceeded, err))
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("if a handler errors with cycle but other handler is falsy, return allowed:false with a nil error", func(t *testing.T) {
+		cyclicErrorHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return nil, ErrCycleDetected
+		}
+
+		resp, err := union(ctx, concurrencyLimit, cyclicErrorHandler, falseHandler)
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+	})
+
+	t.Run("should only aggregate DB queries of truthy handlers", func(t *testing.T) {
+		trueHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: true,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(1),
+				},
+			}, nil
+		}
+
+		falseHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(9),
+				},
+			}, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, trueHandler, falseHandler) // three handlers
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+		require.Equal(t, uint32(9+1), resp.GetResolutionMetadata().DatastoreQueryCount)
+	})
+
+	t.Run("should aggregate DB queries of all falsy handler", func(t *testing.T) {
+		handler := func(context.Context) (*ResolveCheckResponse, error) {
+			return &ResolveCheckResponse{
+				Allowed: false,
+				ResolutionMetadata: &ResolutionMetadata{
+					DatastoreQueryCount: uint32(3),
+				},
+			}, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, handler, handler, handler) // three handlers
+		require.NoError(t, err)
+		require.False(t, resp.GetAllowed())
+		require.Equal(t, uint32(3*3), resp.GetResolutionMetadata().DatastoreQueryCount)
+	})
+
+	t.Run("should return allowed:true if truthy handler evaluated before handler cancels via context", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(ctx)
+
+		cancelHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			time.Sleep(50 * time.Millisecond)
+			cancel()
+			return nil, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, cancelHandler, trueHandler)
+		require.NoError(t, err)
+		require.True(t, resp.GetAllowed())
+	})
+	t.Run("should handle cancellations through context", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(ctx)
+
+		cancelHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			time.Sleep(50 * time.Millisecond)
+			cancel()
+			return nil, nil
+		}
+
+		stuckHandler := func(context.Context) (*ResolveCheckResponse, error) {
+			time.Sleep(1 * time.Hour)
+			return nil, nil
+		}
+
+		resp, err := union(ctx, concurrencyLimit, cancelHandler, stuckHandler)
+		require.ErrorIs(t, err, context.Canceled)
+		require.False(t, resp.GetAllowed())
+	})
+}
```

Comment:
Can we add a test case where we assert what the behavior should be if we see a context deadline and context cancellation prior to seeing an `allowed: true` outcome. In that case we should get an error straight up.